### PR TITLE
[configuration] Support more than one OpenMP implementation

### LINKED ISSUE #256: Support more than one OpenMP implementation
jhunkeler PR comment  

> Perform a generic scan for OpenMP rather than assuming `-fopenmp` is the only flag in town.
> 
> Permits `clang`'s native OpenMP library to link without manual modification of the `wscript` (see: https://openmp.llvm.org/)

### PR DESCRIPTION
**Pull request by jhunkeler**
Script yanked from WAF's documentation:
- https://waf.io/apidocs/_modules/waflib/Tools/fc_config.html

Improvements:
- `./waf configure` now performs a generic scan for OpenMP rather than assuming `-fopenmp` is the only flag in town.
- Permits `clang`'s native OpenMP library to link without manual modification of the `wscript` (see: https://openmp.llvm.org/)
- Makes more sense.

Resolves #256

### CODE REVIEW

**Comment by jamienoss** [comment]
> Permits clang's native OpenMP library to link without manual modification of the wscript (see: https://openmp.llvm.org/)

This doc explicitly states using ``-fopenmp``, which is what it already did. I don't see how this change fa

File: wscript [hunk 1]
```diff
@@ -67,21 +67,27 @@ def options(opt):
 
     opt.recurse('cfitsio')
 
-def _setup_openmp(conf):
-    conf.start_msg('Checking for OpenMP')
-
-    if conf.options.disable_openmp:
```
Comment (jamienoss) [inline wscript]: You forgot to add this section back in.

File: wscript [hunk 2]
```diff
@@ -67,21 +67,27 @@ def options(opt):
...
+    """
+    Detects openmp flags and sets the OPENMP ``CFLAGS``/``LINKFLAGS``
+    """
+    for x in ('-fopenmp','-openmp','-mp','-xopenmp','-omp','-qsmp=omp'):
+        try:
+            conf.check_cc(
+                msg = 'Checking for OpenMP: {0}'.format(x),
+                fragment = '''#include <omp.h>\nint main() { return omp_get_num_threads(); }''',
+                cflags = x,
+                linkflags = (x, '-ldl', '-lpthread', '-lrt'),
```
Comment (jamienoss) [inline wscript]: Why is pthreads needed? Especially since this change is to allow for different implementations of openMP. What happens on Windows?

**Comment by jhunkeler** [comment]
At first glance you'd think so.

Clang's `-fopenmp` flag enables `-lomp` but does not pass the additional libraries  (`dl`, `pthread`, and `rt`) to the linker. The previous implementation requires manual modification of the wscript.

**Comment by jamienoss** [comment]
Before I would got
```
Checking for WAF			:  /Users/jamienoss/dev/hstcal/waf
Setting top to                           : /Users/jamienoss/dev/hstcal 
Setting out to                           : /Users/jamienoss/dev/hstcal/build.Darwin-14.5.0-x86_64-i386-64bit 
Building app                             : HSTCAL 
Version                                  : 1.3.0-33-gcf2a8cef 
git HEAD commit                          : cf2a8cef3cbdde94b0b79c490f0f3dccd2e478dc 
git branch                               : HEAD 
Checking for 'gcc' (C compiler)          : not found 
Checking for 'clang' (C compiler)        : /usr/bin/clang 
Checking for 'gfortran' (Fortran compiler) : /usr/local/bin/gfortran 
Compiling a simple fortran app             : yes 
Checking Mac OS-X version                  : done 
Checking for OpenMP                        : OpenMP not found. 
Checking for sizeof(int)                   : yes 
Checking for compiler flags -O2            : yes 
Checking for compiler flags -Wall          : yes 
Checking for compiler flags -fstack-protector-all : yes 
C compiler flags (CFLAGS)                         : -D APPNAME="HSTCAL" -D VERSION="1.3.0-33-gcf2a8cef" -D BRANCH="HEAD" -D COMMIT="cf2a8cef3cbdde94b0b79c490f0f3dccd2e478dc" -O2 -Wall -fstack-protector-all 
Fortran compiler flags (FCFLAGS)                  : -m64 
Linker flags (LDFLAGS)                            : None 
Using cfitsio                                     : local 
Checking for program 'make'                       : /usr/bin/make 
Copying cfitsio                                   : done 
Configuring cfitsio                               : done 
'configure' finished successfully (6.606s)

```
and now I only get
```
Checking for WAF			:  /Users/jamienoss/dev/hstcal/waf
Setting top to                           : /Users/jamienoss/dev/hstcal 
Setting out to                           : /Users/jamienoss/dev/hstcal/build.Darwin-14.5.0-x86_64-i386-64bit 
Building app                             : HSTCAL 
Version                                  : 1.3.0-34-g745faf50-dirty 
git HEAD commit                          : 745faf500a24af13aecb627f57fec66a32633680 
git branch                               : HEAD 
Checking for 'gcc' (C compiler)          : not found 
Checking for 'clang' (C compiler)        : /usr/bin/clang 
Checking for 'gfortran' (Fortran compiler) : /usr/local/bin/gfortran 
Compiling a simple fortran app             : yes 
Checking Mac OS-X version                  : done 
Checking for OpenMP: -fopenmp              : no 
Checking for OpenMP: -openmp               : no 
Checking for OpenMP: -mp                   : no 
Checking for OpenMP: -xopenmp              : no 
Checking for OpenMP: -omp                  : no 
Checking for OpenMP: -qsmp=omp             : no 
'configure' finished successfully (5.969s)
```
Where's the rest of the output?

**Comment by jamienoss** [comment]
I'm generally apprehensive on the grounds of why isn't this already implemented in [waflib/Tools/c_config.c](https://github.com/waf-project/waf/blob/master/waflib/Tools/c_config.py)?

**Comment by jamienoss** [comment]
> Clang's -fopenmp flag enables -lomp but does not pass the additional libraries (dl, pthread, and rt) to the linker. The previous implementation requires manual modification of the wscript.

But then this would be specific to Clang whereas pthreads would be linked regardless of whether it is needed. Assuming Clang's implementation even uses POSIX threads, does it?

**Comment by jhunkeler** [comment]
I removed the extra link flags and restored the `disable_openmp` flag. If the user happens to know their installation of `clang` has OpenMP built in, they can twiddle `LDFLAGS` on their own accord.

As to why the WAF developers decided to implement `detect_openmp` for only one language: I have no idea. OpenMP clearly isn't specific to Fortran.

**Comment by jamienoss** [comment]
> As to why the WAF developers decided to implement detect_openmp for only one language: I have no idea. OpenMP clearly isn't specific to Fortran.

Maybe there are more implementations in/for fortran and the flags represent this diversity. I think for C/C++ it's just ``-fopenmp``. Do you know explicitly that any of the flags in the list apply to openMP implementations usable by C? I.e. why did you feel that this was needed? E.g. Clang uses ``-fopenmp``.

**Comment by jamienoss** [comment]
What about the now incomplete configure output?
https://github.com/spacetelescope/hstcal/pull/255#issuecomment-336021074

**Comment by jhunkeler** [comment]
1. The incomplete configure output has been taken care of. Using `conf.check_cc` issues its own `end_msg` and completely screws up the message stack. Calling `conf.msg` directly fixes it. The fall-through message "OpenMP not found." was dropped because it is redundant,

2. Does it really matter?

## OpenMP Compilers

Here's a list of all supported compilers and their flags:

http://www.openmp.org/resources/openmp-compilers/

----

## Flag Reference (by documentation)

**GCC:**
From https://gcc.gnu.org/onlinedocs/libgomp/Enabling-OpenMP.html
> `-fopenmp`  must be specified

**Clang:**
From https://openmp.llvm.org

> `-fopenmp` must be linked before it can run. 

**Intel:**
From https://software.intel.com/en-us/node/522690 
> C - `icc -openmp hello.c`
> C++ - `icpc -openmp hello.cpp`
> Fortran - `ifort -openmp hello.f`

There's also a note about explicitly linking to `pthread`:
>When using `gcc` or `g++` compiler to link the application with the Intel® C++ Compiler OpenMP compatibility library, you need to explicitly pass the Intel OpenMP library name using the `-l` option, the Linux* `pthread` library using the `-l` option, and path to the Intel® libraries where the Intel® C++ compiler is installed using the `-L` option

**PGI:**
From  http://www.pgroup.com/resources/docs/17.9/x86/pgi-user-guide/index.htm#openmp-use

> [...] code that can be executed in parallel when the code is compiled using the -⁠mp switch

**IBM XL Fortran:**
From https://www.ibm.com/support/knowledgecenter/en/SSGH4D_15.1.2/com.ibm.xlf151.aix.doc/compiler_ref/opt_smp.html

> The `-qsmp=omp` option must be used to enable OpenMP parallelization.

**Solaris Studio:**
From https://docs.oracle.com/cd/E18659_01/html/821-1381/aewby.html

>If you do not specify `--xopenmp` on the command line, the compiler assumes `--xopenmp=none` (disabling recognition of OpenMP pragmas).

**Visual Studio 20xx:**
(Not implemented here and probably never will be)
From https://msdn.microsoft.com/en-us/library/fw509c3b.aspx

>/openmp (Enable OpenMP 2.0 Support)

**Comment by jamienoss** [comment]
@jhunkeler 
>There's also a note about explicitly linking to pthread:
>
>When using gcc or g++ compiler to link the application with the Intel® C++ Compiler OpenMP compatibility library, you need to explicitly pass the Intel OpenMP library name using the -l option, the Linux* pthread library using the -l option, and path to the Intel® libraries where the Intel® C++ compiler is installed using the -L option

Can we make this conditional for ``-openmp`` then, please? Technically, even that might unnecessarily overlap but...

**Comment by jamienoss** [comment]
@jhunkeler looks good, cheers for making the changes. I'll review it properly later.
FYI We'll want to squash the commits before merging.

**Comment by jamienoss** [comment]
@jhunkeler Hey, if you wonna rebase and squash this, we can merge it.

**Comment by jhunkeler** [comment]
@jamienoss 🚀 ?

**Pull request merged**